cmd/litcli/ln.go

```diff
@@ -237,74 +236,80 @@ type resultStreamWrapper struct {
 //
 // NOTE: This method is part of the PaymentResultStream interface.
 func (w *resultStreamWrapper) Recv() (*lnrpc.Payment, error) {
-	resp, err := w.stream.Recv()
-	if err != nil {
-		return nil, err
-	}
-
-	res := resp.Result
-	switch r := res.(type) {
-	// The very first response might be an accepted sell order, which we
-	// just print out.
-	case *tchrpc.SendPaymentResponse_AcceptedSellOrder:
-		quote := r.AcceptedSellOrder
+	// printQuote unmarshals and prints an accepted quote.
+	printQuote := func(quote *rfqrpc.PeerAcceptedSellQuote) error {
 		rpcRate := quote.BidAssetRate
 		rate, err := rpcutils.UnmarshalRfqFixedPoint(rpcRate)
 		if err != nil {
-			return nil, fmt.Errorf("unable to unmarshal fixed "+
-				"point: %w", err)
+			return fmt.Errorf("unable to unmarshal fixed point: %w",
+				err)
 		}
 
 		amountMsat := lnwire.MilliSatoshi(w.amountMsat)
 		milliSatsFP := rfqmath.MilliSatoshiToUnits(amountMsat, *rate)
 		numUnits := milliSatsFP.ScaleTo(0).ToUint64()
 
-		// If the calculated number of units is 0 then the asset rate
-		// was not sufficient to represent the value of this payment.
+		// The purpose of this function is just to print, so let's avoid
+		// dividing by zero or reporting an invalid msat/unit rate.
 		if numUnits == 0 {
-			// We will calculate the minimum amount that can be
-			// effectively sent with this asset by calculating the
-			// value of a single asset unit, based on the provided
-			// asset rate.
-
-			// We create the single unit.
-			unit := rfqmath.FixedPointFromUint64[rfqmath.BigInt](
-				1, 0,
-			)
-
-			// We derive the minimum amount.
-			minAmt := rfqmath.UnitsToMilliSatoshi(unit, *rate)
-
-			// We return the error to the user.
-			return nil, fmt.Errorf("smallest payment with asset "+
-				"rate %v is %v, cannot send %v",
-				rate.ToUint64(), minAmt, amountMsat)
+			return nil
 		}
 
 		msatPerUnit := uint64(w.amountMsat) / numUnits
 
 		fmt.Printf("Got quote for %v asset units at %v msat/unit from "+
-			"peer %s with SCID %d\n", numUnits, msatPerUnit,
+			" peer %s with SCID %d\n", numUnits, msatPerUnit,
 			quote.Peer, quote.Scid)
 
-		resp, err = w.stream.Recv()
+		return nil
+	}
+
+	// A boolean to indicate whether the first quote was printed via the
+	// legacy single-rfq response field.
+	legacyFirstPrint := false
+
+	for {
+		resp, err := w.stream.Recv()
 		if err != nil {
 			return nil, err
 		}
 
-		if resp == nil || resp.Result == nil ||
-			resp.GetPaymentResult() == nil {
+		res := resp.Result
 
-			return nil, errors.New("unexpected nil result")
-		}
+		switch r := res.(type) {
```

Comment:
Switching over `type` feels strange to me but I have little context on `litd`.
I'm surprised we need to differentiate between deprecated `AcceptSellOrder` and the new `AcceptedSellOrders`. If the `litd` version supporting multi-rfq is released after the `tapd` version that adds the new `AcceptedSellOrders`, it seems reasonable that `litd` should simply not support the deprecated type. This is backwards compatibility squared.

Otherwise might be worth adding a comment explaining when the support for legacy can be dropped  because there is no direct correlation between dropping support in `tapd` and `litd`.

Reply:
Agree that switching over the `type` feels a bit strange, but it's commonly used throughout our daemons, so totally ok IMO.

Good point in regards to the supporting of the deprecated type! Something to note is that for our other cli's we do allow `remote` mode, and can therefore not know exactly which `daemon` version the user is running. But for `tapd`, remote mode is not yet supported.

Reply:
Some more context on the new RPC field / deprecation of the old one:
https://github.com/lightninglabs/taproot-assets/pull/1613#discussion_r2224642966